ISSUE-155: Adds JSON Patch Action example (#162)
esmero/archipelago-documentation: docs/find_and_replace_action_json_patch.md

    } 
]
```

 - `/subject_lcgft_terms/0` 
 
```JSON
{
    "uri": "http:\/\/id.loc.gov\/authorities\/genreForms\/gf2017027249",
    "label": "Photographs"
}
```

-  `/subject_lcgft_terms/0/label`

```
Photographs
```

AS you can see a JSON Pointer is very precise , allowing you to target complete structures and values but it *does not allow* Wildcard Operations. Means you can not "search" or do loosy comparissons. This very fact limits many times the use case. E.g. if you have a list of terms like this:

```JSON
"terms": [
	"term1",
	"term2",
	"term3"
]
```

and you want to "test" for the existence of `"term3"` before applying a change, you would need to know exactly at what position inside the `terms Array` (Starting from 0) it will/should be found.  And that might not be consistent across every ADO. 

So how do we use these pointers in an operation inside a JSON Patch document? Using the same fake "terms" JSON snippet the previously listed, operations are written like this:

## add

```JSON
{ "op": "add", "path": "/terms/0", "value": "term_another" }
```

This will add before the first term (in this case "term1" ) "term_another" as a value.

??? info "At the end"

> 	you can use a dash (`-`) (e.g. "/terms/-") instead of the numeric position of an array entry to denote that the "value" needs to be added at the end. This is needed specially for empty lists. You can not target `0` position on an empty array.

Reviewer: ```suggestion
    you can use a dash (`-`) (e.g. "/terms/-") instead of the numeric position of an array entry to denote that the "value" needs to be added at the end. This is needed specially for empty lists. You can not target `0` position on an empty array.
```

As with every other VBO action described in our documentation, start by selecting the ADOs you want to modify using the exposed Search Field(s) and/or Facets present in the Search and Replace `View` found at `/search-and-replace` . 

Once you have filtered down the list to manageable size, containing at least the ADOs you plan on modifying (but for JSON Patch operation could be more too because you can “test“ and match ), press either on `Select / deselect all results in this view` to pass all  the result (this includes all pages, not only the currently visible one) or go selectively one by one by checking on the `toggle` found beside each ADO's title. You will see how the number in `Selected 0 item in this view` increases. Now press `Apply to select Items`.  We will use for this example `Laddie the dog running in the garden, Bronx, N.Y., undated [c. 1910-1918?]`, an ADO we provide in our DEMO sets. 

Redundant to say, Batch Actions are intended to be used when a modification needs to be applied to more than one item, implying there is a pattern. For single ADOs you can always do this faster directly via the **EDIT** tab.

??? note "Keep your JSON Patches (and friends) around"

> 	Since JSON Patching involves writing a, sometimes complex, JSON Document, please keep around an Application (or Text File) where you can copy/paste and save your JSON Patches for resuse or future references. Archipelago will not store nor remember between runs the JSON Patch document you submitted. It is also very useful to copy and have at hand the `RAW JSON` of one of the ADOs you plan to modify as a reference/aid while building the given JSON Patch document.

Reviewer: ```suggestion
    Since JSON Patching involves writing a, sometimes complex, JSON Document, please keep around an Application (or Text File) where you can copy/paste and save your JSON Patches for resuse or future references. Archipelago will not store nor remember between runs the JSON Patch document you submitted. It is also very useful to copy and have at hand the `RAW JSON` of one of the ADOs you plan to modify as a reference/aid while building the given JSON Patch document.
```

`
## Step 2: JSON Patch Metadata for Archipelago Digital Objects Action Configuration

The default config JSON Patch form will contain an example JSON Patch set of Commands (Document)

![Default JSON Patch Command](images/json_patch_config_default.jpg)

We are going to replace this one with a valid JSON document. Notice that it does not require a root `{}` Object wrapper and it is really a list (or array) of operations. 

A bit of repetition but needed to explain the JSON Patch document. You can see on the following Note box how we copyed the RAW JSON of one ADO to be Patched to have a reference while building the JSON PATCH. for the sake of brevity we remove here the longer Image File technical Metadata.

> ??? info "RAW JSON of `Laddie the dog running in the garden, Bronx, N.Y., undated [c. 1910-1918?]``"

Reviewer: ```suggestion
??? info "RAW JSON of `Laddie the dog running in the garden, Bronx, N.Y., undated [c. 1910-1918?]`"
```

3. And the extra LOD entry for wikidata, in this case the Q Item for [collie](https://www.wikidata.org/wiki/Q1196071) (type of herding dog)

```JSON
{ "op": "add", "path": "/subject_wikidata/-", "value": {
        "uri": "https://www.wikidata.org/wiki/Q1196071",
        "label": "collie"
    }
}
```

4. Finally we are going to copy the `geographic_location.state`  value and put it into `subjects_local`:

```JSON
{"op": "remove", "path": "/subjects_local"},
{"op": "add", "path": "/subjects_local", "value": []},
{"op": "copy", "from": "/geographic_location/state", "path": "/subjects_local/-"}
```

Why so many operations? Because initially ` "subjects_local"` had a value of `null` and it is not suited to generate/add to it as a multivalued key because of that.  So we need to remove it first, recreate it as an empty list and then we can copy. **Pro Note:** you could partially rewrite this as  `replace` operation!

??? info "what if it subjects_local has data?

> 	You will lose it! you can add a “test“ or move data instead of recreating. Many choices.

Reviewer: ```suggestion
    You will lose it! you can add a “test“ or move data instead of recreating. Many choices.
```

The final JSON Patch will look like this. Copy it into the Configuration JSON Patch commands form field: 

```JSON
[
    {"op": "test", "path": "/type", "value": "Photograph"},
    {"op": "test", "path": "/ismemberof", "value": [16]},
    {"op": "add","path": "/temp_description_array","value": []},
    {"op": "move","from": "/description","path": "/temp_description_array/-"},
    {"op": "move","from": "/temp_description_array","path": "/description"},
    {"op": "add","path": "/subject_wikidata/-","value": 
        {
        "uri": "https://www.wikidata.org/wiki/Q1196071",
         "label": "collie"
         }
    },
    {"op": "remove", "path": "/subjects_local"},
    {"op": "add", "path": "/subjects_local", "value": []},
    {"op": "copy", "from": "/geographic_location/state", "path": "/subjects_local/-"}
]
```

??? note "The inverse process online"

> 	Now that you know (or are starting to understand) the manual process you can also try this [online tool](https://json-patch-builder-online.github.io) that allows you, based on a source and a destination JSON, generate the needed JSON Patch to mutate one JSON into the another. The logic might not be always what you need and most likely it will not take in account that you actually need to move values and will prefer to fix values to be added via an add operation

Reviewer: ```suggestion
    Now that you know (or are starting to understand) the manual process you can also try this [online tool](https://json-patch-builder-online.github.io) that allows you, based on a source and a destination JSON, generate the needed JSON Patch to mutate one JSON into the another. The logic might not be always what you need and most likely it will not take in account that you actually need to move values and will prefer to fix values to be added via an add operation
```

        }
    ],
    "upload_associated_warcs": [],
    "physical_description_extent": "",
    "subject_lcnaf_personal_names": "",
    "subject_lcnaf_corporate_names": "",
    "subjects_local_personal_names": "",
    "related_item_host_location_url": null,
    "subject_lcnaf_geographic_names": [
        {
            "uri": "http:\/\/id.loc.gov\/authorities\/names\/n81059724",
            "label": "Bronx (New York, N.Y.)"
        },
        {
            "uri": "http:\/\/id.loc.gov\/authorities\/names\/n79007751",
            "label": "New York (N.Y.)"
        }
    ],
    "related_item_host_display_label": null,
    "related_item_host_local_identifier": null,
    "related_item_host_title_info_title": "",
    "related_item_host_type_of_resource": null,
    "physical_description_note_condition": null
}
```

That is all. Again, keep your **JSON Patches** safe in a text document, test/try simple things first, look for **patterns**, look for *No-Nos* that can become "tests" to avoid touching ADOs that do not need to be updated and always remember that the destination type (single value, array or object) of an existing Key might affect your complex logic. Happy Patching!

> 

Reviewer: ```suggestion

___

```

A JSON Patch is a `JSON` Document containing precise operations that can heavily modify the structure and values of an existing JSON Document, in this case the RAW JSON found inside a strawberryfield of our ADOs. 

The operations  available for modifications of an JSON document are:

- “add”
- “remove”
- “replace”,
- “move”
- “copy”

And there is also one (very important) used to check/validate the existence of values/keys:

- “test”

Even if you can have multiple operations in a single JSON Patch Document, they are always applied as a whole. Means if any of those fail nothing will be applied and in concrete, in our VBO action, no change will be done to your ADO. This in combination with the “test” operation gives you a lot of safety and a way of discerning/skipping completely a complex set of operations on e.g. a failed ”test”.

JSON Patch uses `JSON Pointers` as arguments in all of these operations to target a specific key/value of your JSON.

Using the following JSON Snippet as example:

```JSON
"subject_lcgft_terms": [
        {
            "uri": "http:\/\/id.loc.gov\/authorities\/genreForms\/gf2017027249",
            "label": "Photographs"

>         }

Reviewer 2: ```suggestion
    {
        "uri": "http:\/\/id.loc.gov\/authorities\/genreForms\/gf2017027249",
        "label": "Photographs"
    }
```
Normalize spacing.

]
```

These JSON Pointers will resolve in the following manner:

- `/subject_lcgft_terms` 

```JSON
[

>     {
>             "uri": "http:\/\/id.loc.gov\/authorities\/genreForms\/gf2017027249",
>             "label": "Photographs"
>     } 

Reviewer 2: ```suggestion
    {
        "uri": "http:\/\/id.loc.gov\/authorities\/genreForms\/gf2017027249",
        "label": "Photographs"
    }
```
Normalize spacing.

]
```

>  - `/subject_lcgft_terms/0` 

Reviewer 2: ```suggestion
- `/subject_lcgft_terms/0`
```
Stray spaces.

## add

```JSON
{ "op": "add", "path": "/terms/0", "value": "term_another" }
```

This will add before the first term (in this case "term1" ) "term_another" as a value.

??? info "At the end"

    you can use a dash (`-`) (e.g. "/terms/-") instead of the numeric position of an array entry to denote that the "value" needs to be added at the end. This is needed specially for empty lists. You can not target `0` position on an empty array.

## remove

```JSON
{ "op": "remove", "path": "/terms/1"}
```

This will remove the second term  (in this case "term2" )

## replace

```JSON
{ "op": "replace", "path": "/terms/1", "value": "term_again" }
```
This will remove the second term  (in this case "term2" ) and put in its place "term_again" as a value. Basically two operationes, “remove“ and “add “ in a single step

## move

```JSON
{ "op": "move", "from": "/terms", "path": "/terms_somewhere_else"}
```

This will copy  all values inside top JSON key `terms` into a new top JSON key named `terms_somewhere_else` and remove then the old `terms` key.

## copy

```JSON
{ "op": "copy", "from": "/terms", "path": "/terms_somewhere_else"}
```

> Similar to “move“, it will copylvalues inside top JSON key `terms` into a new top JSON key named `terms_somewhere_else` without removing then the old `terms` key or its content!

Reviewer 2: ```suggestion
Similar to “move“, it will copy values inside top JSON key `terms` into a new top JSON key named `terms_somewhere_else` without removing then the old `terms` key or its content!
```
Typo.
